Support data validation using value and tolerance

danielhuppmann opened the pull request:
This PR implements support for configuring validation criteria using value and atol/rtol (optional). It also adds some safeguards that only either bounds or value (and tolerance) can be configured in a **DataValidationCriteria** instance. And adds a few tests.

phackstock reviewed (commented):
As just discussed bilaterally, I think it would be a good idea to subclass `DataValidationCriteria`.
You'd make two classes `UpperLowerCriteria(DataValidationCriteria)` and `ValuePlusToleranceCriteria(DataValidationCriteria)` (or some better name that what I can come up with now).
`UpperLowerCriteria` would be mostly what `DataValidationCritera` is now with attributes `upper_bound` and `lower_bound`. 
`ValuePlusToleranceCriteria(DataValidationCriteria)` would have `value`, `rtol`, and `atol`. It would also implement two properties `upper_bound` and `lower_bound`. 
This way both classes can be used in the same way when validating data.

danielhuppmann commented on nomenclature/processor/data_validator.py:
```diff
@@ -21,6 +23,29 @@ class DataValidationCriteria(IamcDataFilter):
+    rtol: float = None
+    atol: float = None
+
+    @model_validator(mode="before")
```
No, seems to be required.

danielhuppmann commented:
Thanks @phackstock, implemented the optional-field and don't allow extra fields anymore.

For separating this out into separate classes for bounds and value/tolerance, the issue is that this yields a very user-unfriendly error message if a validation-item has both value and bounds. I implemented this on the branch https://github.com/danielhuppmann/nomenclature/tree/feature/validate-range-subclasses, maybe you have an idea on how to generate user-intelligible error messages?

phackstock commented:
@danielhuppmann, just took a look at your extra branch, implementation wise that's exactly what I meant. 
I'll see if I can figure out a neat way to make the error messages better for the user.

danielhuppmann commented:
closing in favor of #371
